pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go

```diff
@@ -278,135 +299,237 @@ func CreatePoller(
 	qemuAgentVersionInterval time.Duration,
 	qemuAgentFSFreezeStatusInterval time.Duration,
 ) *AgentPoller {
-	p := &AgentPoller{
-		Connection: connecton,
+	return &AgentPoller{
+		Connection: connection,
 		VmiUID:     vmiUID,
 		domainName: domainName,
 		agentStore: store,
-		workers:    []PollerWorker{},
-	}
-
-	// version command group
-	p.workers = append(p.workers, PollerWorker{
-		CallTick:      qemuAgentVersionInterval,
-		AgentCommands: []AgentCommand{GET_AGENT},
-	})
-	// sys command group
-	p.workers = append(p.workers, PollerWorker{
-		CallTick:      qemuAgentSysInterval,
-		AgentCommands: []AgentCommand{GET_INTERFACES, GET_OSINFO, GET_TIMEZONE, GET_HOSTNAME},
-	})
-	// filesystem command group
-	p.workers = append(p.workers, PollerWorker{
-		CallTick:      qemuAgentFileInterval,
-		AgentCommands: []AgentCommand{GET_FILESYSTEM},
-	})
-	// user command group
-	p.workers = append(p.workers, PollerWorker{
-		CallTick:      qemuAgentUserInterval,
-		AgentCommands: []AgentCommand{GET_USERS},
-	})
-	// fsfreeze command group
-	p.workers = append(p.workers, PollerWorker{
-		CallTick:      qemuAgentFSFreezeStatusInterval,
-		AgentCommands: []AgentCommand{GET_FSFREEZE_STATUS},
-	})
-
-	return p
+		workers: []PollerWorker{
+			// Polling for QEMU agent commands
+			{
+				CallTick:      qemuAgentVersionInterval,
+				AgentCommands: []AgentCommand{GetAgent},
+			},
+			{
+				CallTick:      qemuAgentFileInterval,
+				AgentCommands: []AgentCommand{GetFilesystem},
+			},
+			{
+				CallTick:      qemuAgentFSFreezeStatusInterval,
+				AgentCommands: []AgentCommand{GetFSFreezeStatus},
+			},
+			// Polling for guest info API
+			{
+				CallTick:  qemuAgentSysInterval,
+				InfoTypes: INTERFACES | OSINFO | HOSTNAME | TIMEZONE,
+			},
+			{
+				CallTick:  qemuAgentUserInterval,
+				InfoTypes: USERS,
+			},
+		},
+	}
 }
 
-// Start the poller workers
+// Start the poller workers and libvirt API operations
 func (p *AgentPoller) Start() {
 	if p.agentDone != nil {
 		return
 	}
 	p.agentDone = make(chan struct{})
 
-	for i := 0; i < len(p.workers); i++ {
-		log.Log.Infof("Starting agent poller with commands: %v", p.workers[i].AgentCommands)
-		go p.workers[i].Poll(func(commands []AgentCommand) {
-			executeAgentCommands(commands, p.Connection, p.agentStore, p.domainName)
-		}, p.agentDone, pollInitialInterval)
+	for _, worker := range p.workers {
+		if len(worker.AgentCommands) != 0 {
+			log.Log.Infof("Starting agent poller with commands: %v", worker.AgentCommands)
+		} else {
+			log.Log.Infof("Starting agent poller with API operations: %v", worker.InfoTypes)
+		}
```

Comment:
In theory it could do both

Reply:
I'd prefer to separate it like we do the execution of it separately.